pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -2784,32 +2786,55 @@ public String startReplaceSegments(String tableNameWithType, List<String> segmen
         Preconditions.checkArgument(segmentLineage.getLineageEntry(segmentLineageEntryId) == null,
             String.format("SegmentLineageEntryId (%s) already exists in the segment lineage.", segmentLineageEntryId));
 
+        List<String> segmentsToCleanUp = new ArrayList<>();
         for (String entryId : segmentLineage.getLineageEntryIds()) {
           LineageEntry lineageEntry = segmentLineage.getLineageEntry(entryId);
 
-          // If segment entry is in 'REVERTED' state, no need to check for 'segmentsFrom'.
-          if (lineageEntry.getState() != LineageEntryState.REVERTED) {
+          // If the lineage entry is in 'REVERTED' state, no need to go through the validation because we can regard
+          // the entry as not existing.
+          if (lineageEntry.getState() == LineageEntryState.REVERTED) {
+            continue;
+          }
+
+          // By here, the lineage entry is either 'IN_PROGRESS' or 'COMPLETED'.
+          if (forceCleanup && lineageEntry.getState() == LineageEntryState.IN_PROGRESS && CollectionUtils
+              .isEqualCollection(segmentsFrom, lineageEntry.getSegmentsFrom())) {
+            // When 'forceCleanup' is enabled, we need to proactively revert the lineage entry when we find the lineage
+            // entry with the same 'segmentFrom' values.
+            segmentLineage.updateLineageEntry(entryId,
```

Comment:
We need to check if the `segmentsFrom` are online before updating the status as `REVERTED`.

Reply:
added a function for `updateSegmentLineageEntryToReverted()` and reuse the same function from both `startReplaceSegments` and `revertReplaceSegments`

Reply:
LGTM